shell/platform/embedder/embedder_external_view_embedder.cc

```diff
@@ -0,0 +1,223 @@
+  FlutterLayer layer = {};
+
+  layer.struct_size = sizeof(layer);
+  layer.type = kFlutterLayerContentTypeBackingStore;
+  layer.backing_store = store;
+
+  CopyViewParams(layer, params);
+
+  return layer;
+}
+
+static FlutterPlatformView MakePlatformView(
+    FlutterPlatformViewIdentifier identifier) {
+  FlutterPlatformView view = {};
+
+  view.struct_size = sizeof(view);
+
+  view.identifier = identifier;
+
+  return view;
+}
+
+static FlutterLayer MakeLayer(const EmbeddedViewParams& params,
+                              const FlutterPlatformView& platform_view) {
+  FlutterLayer layer = {};
+
+  layer.struct_size = sizeof(layer);
+  layer.type = kFlutterLayerContentTypePlatformView;
+  layer.platform_view = &platform_view;
+
+  CopyViewParams(layer, params);
+
+  return layer;
+}
+
+// |ExternalViewEmbedder|
+bool EmbedderExternalViewEmbedder::SubmitFrame(GrContext* context) {
+  std::map<FlutterPlatformViewIdentifier, FlutterPlatformView>
+      presented_platform_views;
+  // Layers may contain pointers to platform views in the collection above.
+  std::vector<FlutterLayer> presented_layers;
+  Registry render_targets_used;
+
+  for (const auto& view_id : composition_order_) {
+    FML_DCHECK(pending_recorders_.count(view_id) == 1);
+    FML_DCHECK(pending_params_.count(view_id) == 1);
+
+    const auto& params = pending_params_.at(view_id);
+    auto& recorder = pending_recorders_.at(view_id);
+
+    auto picture = recorder->finishRecordingAsPicture();
+    if (!picture) {
+      FML_LOG(ERROR) << "Could not finish recording into the picture before "
+                        "on-screen composition.";
+      return false;
+    }
+
+    const auto backing_store_config =
+        MakeBackingStoreConfig(pending_frame_size_);
+
+    RegistryKey registry_key(view_id, backing_store_config);
+
+    auto found_render_target = registry_.find(registry_key);
+
+    // Find a cached render target in the registry. If none exists, ask the
+    // embedder for a new one.
+    std::shared_ptr<EmbedderRenderTarget> render_target;
+    if (found_render_target == registry_.end()) {
+      render_target = render_target_callback_(context, backing_store_config);
+    } else {
+      render_target = found_render_target->second;
+    }
+
+    if (!render_target) {
+      FML_LOG(ERROR) << "Could not acquire external render target for "
+                        "on-screen composition.";
+      return false;
+    }
+
+    render_targets_used[registry_key] = render_target;
+
+    auto render_canvas = render_target->GetRenderCanvas();
+
+    if (!render_canvas) {
+      FML_LOG(ERROR)
+          << "Could not acquire render canvas for on-screen rendering.";
+      return false;
+    }
+
+    render_canvas->clear(SK_ColorTRANSPARENT);
+    render_canvas->drawPicture(picture);
+    render_canvas->flush();
+
+    // Indicate a layer for the platform view.
+    presented_platform_views[view_id] = MakePlatformView(view_id);
```

Comment:
What is `presented_platform_views` used for?

Reply:
`presented_layers` contain pointers to objects in `presented_platform_views` (see the comment on line 144). These structs must be alive till the `present_callback_` is invoked.

Reply:
That is why I added the comment on line 165. Were you referring to something else?

Reply:
I guess that comment wasn't explicit enough for me 😄 
I'd suggest adding a comment here, saying that we add to presented_platform_views in order to keep at allocated just for the scope of the current method.

Reply:
Done.